lxd/patches.go

```diff
@@ -769,4 +770,81 @@ func patchNetworkClearBridgeVolatileHwaddr(name string, d *Daemon) error {
 	return nil
 }
 
+func patchZfsDropBlockVolumeFilesystemExtension(name string, d *Daemon) error {
+	s := d.State()
+
+	poolNames, err := s.DB.Cluster.GetStoragePoolNames()
+	if err != nil {
+		return fmt.Errorf("Failed getting storage pool names: %w", err)
+	}
+
+	var zfsPoolNames []string
+
+	for _, poolName := range poolNames {
+		err := s.DB.Cluster.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+			poolID, err := tx.GetStoragePoolID(ctx, poolName)
+			if err != nil {
+				return err
+			}
+
+			driverName, err := tx.GetStoragePoolDriver(ctx, poolID)
+			if err != nil {
+				return err
+			}
+
+			if driverName != "zfs" {
+				return nil
+			}
+
+			configs, err := tx.GetStoragePoolsLocalConfig(ctx)
+			if err != nil {
+				return err
+			}
+
+			// Get the real zfs pool name.
+			zfsPoolName, ok := configs[poolName]["zfs.pool_name"]
+			if !ok {
+				zfsPoolName = poolName
+			}
+
+			zfsPoolNames = append(zfsPoolNames, zfsPoolName)
+
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, poolName := range zfsPoolNames {
+		out, err := shared.RunCommand("zfs", "list", "-H", "-r", "-o", "name", "-t", "volume", fmt.Sprintf("%s/images", poolName))
+		if err != nil {
+			return fmt.Errorf("Failed listing images: %w", err)
+		}
+
+		for _, volume := range strings.Split(out, "\n") {
+			fields := strings.SplitN(volume, "/", 3)
+
+			if len(fields) != 3 {
+				continue
+			}
+
+			// Ignore non-block images, and images without filesystem extension
+			if !strings.HasSuffix(fields[2], ".block") || !strings.Contains(fields[2], "_") {
+				continue
+			}
+
+			// Rename zfs dataset. Snapshots will automatically be renamed.
+			newName := fmt.Sprintf("%s/images/%s.block", poolName, strings.Split(fields[2], "_")[0])
+
+			_, err = shared.RunCommand("zfs", "rename", volume, newName)
```

Comment:
@monstermunchkin did you test this when reloading LXD with a running instance that is using a volume named in the old style? (as this is what snap refresh will do).

Reply:
> did you test this when reloading LXD with a running instance that is using a volume named in the old style?

Not yet. 

Also, since I have a better understanding of the patches now, I'll change it so that it uses the `patchGenericStorage` function.

Reply:
So, nothing happens when renaming the image block volume while a VM is running.

Reply:
As in it gets renamed in the background but doesn't affect the VM running?

Reply:
Only the image is renamed. The VM is a separate dataset/volume.

Reply:
Ah right ofcourse, the PR title made me think this was for VM volumes, but ofcourse, the issue is with the VM image volumes.